// @allurereport/core/dist/utils/event.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _RealtimeEventsDispatcher_emitter, _RealtimeSubscriber_instances, _RealtimeSubscriber_emitter, _RealtimeSubscriber_handlers, _RealtimeSubscriber_onEvent, _RealtimeSubscriber_onBatchedEvent, _RealtimeSubscriber_createBatchHandler, _RealtimeSubscriber_runBatchHandler, _RealtimeSubscriber_disposeBatchHandler;
import console from "node:console";
import { setTimeout } from "node:timers/promises";
export var RealtimeEvents;
(function (RealtimeEvents) {
    RealtimeEvents["TestResult"] = "testResult";
    RealtimeEvents["TestFixtureResult"] = "testFixtureResult";
    RealtimeEvents["AttachmentFile"] = "attachmentFile";
    RealtimeEvents["QualityGateResults"] = "qualityGateResults";
    RealtimeEvents["GlobalAttachment"] = "globalAttachment";
    RealtimeEvents["GlobalError"] = "globalError";
    RealtimeEvents["GlobalExitCode"] = "globalExitCode";
})(RealtimeEvents || (RealtimeEvents = {}));
const runListener = (listener, ...args) => {
    try {
        return Promise.resolve(listener(...args)).catch((err) => {
            console.error("can't execute listener", err);
        });
    }
    catch (err) {
        console.error("can't execute listener", err);
        return Promise.resolve();
    }
};
const createListenerHandler = (listener) => (...args) => {
    void runListener(listener, ...args);
};
export class RealtimeEventsDispatcher {
    constructor(emitter) {
        _RealtimeEventsDispatcher_emitter.set(this, void 0);
        __classPrivateFieldSet(this, _RealtimeEventsDispatcher_emitter, emitter, "f");
    }
    sendGlobalAttachment(attachment, fileName, environment) {
        __classPrivateFieldGet(this, _RealtimeEventsDispatcher_emitter, "f").emit(RealtimeEvents.GlobalAttachment, { attachment, fileName, environment });
    }
    sendGlobalExitCode(codes) {
        __classPrivateFieldGet(this, _RealtimeEventsDispatcher_emitter, "f").emit(RealtimeEvents.GlobalExitCode, codes);
    }
    sendGlobalError(error) {
        __classPrivateFieldGet(this, _RealtimeEventsDispatcher_emitter, "f").emit(RealtimeEvents.GlobalError, error);
    }
    sendQualityGateResults(payload) {
        __classPrivateFieldGet(this, _RealtimeEventsDispatcher_emitter, "f").emit(RealtimeEvents.QualityGateResults, payload ?? []);
    }
    sendTestResult(trId) {
        __classPrivateFieldGet(this, _RealtimeEventsDispatcher_emitter, "f").emit(RealtimeEvents.TestResult, trId);
    }
    sendTestFixtureResult(tfrId) {
        __classPrivateFieldGet(this, _RealtimeEventsDispatcher_emitter, "f").emit(RealtimeEvents.TestFixtureResult, tfrId);
    }
    sendAttachmentFile(afId) {
        __classPrivateFieldGet(this, _RealtimeEventsDispatcher_emitter, "f").emit(RealtimeEvents.AttachmentFile, afId);
    }
}
_RealtimeEventsDispatcher_emitter = new WeakMap();
export class RealtimeSubscriber {
    constructor(emitter) {
        _RealtimeSubscriber_instances.add(this);
        _RealtimeSubscriber_emitter.set(this, void 0);
        _RealtimeSubscriber_handlers.set(this, []);
        __classPrivateFieldSet(this, _RealtimeSubscriber_emitter, emitter, "f");
    }
    onGlobalAttachment(listener) {
        return __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_onEvent).call(this, RealtimeEvents.GlobalAttachment, listener);
    }
    onGlobalExitCode(listener) {
        return __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_onEvent).call(this, RealtimeEvents.GlobalExitCode, listener);
    }
    onGlobalError(listener) {
        return __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_onEvent).call(this, RealtimeEvents.GlobalError, listener);
    }
    onQualityGateResults(listener) {
        return __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_onEvent).call(this, RealtimeEvents.QualityGateResults, listener);
    }
    onTestResults(listener, options = {}) {
        return __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_onBatchedEvent).call(this, RealtimeEvents.TestResult, listener, options);
    }
    onTestFixtureResults(listener, options = {}) {
        return __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_onBatchedEvent).call(this, RealtimeEvents.TestFixtureResult, listener, options);
    }
    onAttachmentFiles(listener, options = {}) {
        return __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_onBatchedEvent).call(this, RealtimeEvents.AttachmentFile, listener, options);
    }
    offAll() {
        __classPrivateFieldGet(this, _RealtimeSubscriber_emitter, "f").removeAllListeners();
        for (const handler of __classPrivateFieldGet(this, _RealtimeSubscriber_handlers, "f")) {
            __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_disposeBatchHandler).call(this, handler);
        }
        __classPrivateFieldSet(this, _RealtimeSubscriber_handlers, [], "f");
    }
}
_RealtimeSubscriber_emitter = new WeakMap(), _RealtimeSubscriber_handlers = new WeakMap(), _RealtimeSubscriber_instances = new WeakSet(), _RealtimeSubscriber_onEvent = function _RealtimeSubscriber_onEvent(event, listener) {
    const handler = createListenerHandler(listener);
    __classPrivateFieldGet(this, _RealtimeSubscriber_emitter, "f").on(event, handler);
    return () => {
        __classPrivateFieldGet(this, _RealtimeSubscriber_emitter, "f").off(event, handler);
    };
}, _RealtimeSubscriber_onBatchedEvent = function _RealtimeSubscriber_onBatchedEvent(event, listener, options) {
    const { maxTimeout = 100 } = options;
    const { dispose, eventHandler } = __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_createBatchHandler).call(this, maxTimeout, listener);
    __classPrivateFieldGet(this, _RealtimeSubscriber_emitter, "f").on(event, eventHandler);
    return () => {
        __classPrivateFieldGet(this, _RealtimeSubscriber_emitter, "f").off(event, eventHandler);
        dispose();
    };
}, _RealtimeSubscriber_createBatchHandler = function _RealtimeSubscriber_createBatchHandler(maxTimeout, listener) {
    const handler = {
        buffer: [],
        closed: false,
    };
    __classPrivateFieldGet(this, _RealtimeSubscriber_handlers, "f").push(handler);
    const eventHandler = (trId) => {
        if (handler.closed) {
            return;
        }
        handler.buffer.push(trId);
        if (handler.cycle) {
            return;
        }
        handler.cycle = __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_runBatchHandler).call(this, handler, maxTimeout, listener);
    };
    return {
        dispose: () => __classPrivateFieldGet(this, _RealtimeSubscriber_instances, "m", _RealtimeSubscriber_disposeBatchHandler).call(this, handler),
        eventHandler,
    };
}, _RealtimeSubscriber_runBatchHandler = async function _RealtimeSubscriber_runBatchHandler(handler, maxTimeout, listener) {
    try {
        while (!handler.closed && handler.buffer.length > 0) {
            handler.abortController = new AbortController();
            await setTimeout(maxTimeout, undefined, { signal: handler.abortController.signal });
            handler.abortController = undefined;
            const bufferCopy = [...handler.buffer];
            handler.buffer = [];
            await runListener(listener, bufferCopy);
        }
    }
    catch (err) {
        if (err.name === "AbortError") {
            return;
        }
        handler.buffer = [];
        console.error("can't execute listener", err);
    }
    finally {
        handler.abortController = undefined;
        handler.cycle = undefined;
    }
}, _RealtimeSubscriber_disposeBatchHandler = function _RealtimeSubscriber_disposeBatchHandler(handler) {
    handler.closed = true;
    handler.abortController?.abort();
    handler.abortController = undefined;
    handler.buffer = [];
    handler.cycle = undefined;
    __classPrivateFieldSet(this, _RealtimeSubscriber_handlers, __classPrivateFieldGet(this, _RealtimeSubscriber_handlers, "f").filter((registeredHandler) => registeredHandler !== handler), "f");
};
