// @allurereport/core/dist/utils/realtimeChannel.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _RealtimeChannel_emitter;
import { EventEmitter } from "node:events";
import { RealtimeEvents, RealtimeEventsDispatcher, RealtimeSubscriber } from "./event.js";
export class RealtimeChannel {
    constructor() {
        _RealtimeChannel_emitter.set(this, new EventEmitter());
        this.dispatcher = new RealtimeEventsDispatcher(__classPrivateFieldGet(this, _RealtimeChannel_emitter, "f"));
        this.subscriber = new RealtimeSubscriber(__classPrivateFieldGet(this, _RealtimeChannel_emitter, "f"));
    }
    onResultLikeChanged(listener) {
        __classPrivateFieldGet(this, _RealtimeChannel_emitter, "f").on(RealtimeEvents.TestResult, listener);
        __classPrivateFieldGet(this, _RealtimeChannel_emitter, "f").on(RealtimeEvents.TestFixtureResult, listener);
        __classPrivateFieldGet(this, _RealtimeChannel_emitter, "f").on(RealtimeEvents.AttachmentFile, listener);
        return () => {
            __classPrivateFieldGet(this, _RealtimeChannel_emitter, "f").off(RealtimeEvents.TestResult, listener);
            __classPrivateFieldGet(this, _RealtimeChannel_emitter, "f").off(RealtimeEvents.TestFixtureResult, listener);
            __classPrivateFieldGet(this, _RealtimeChannel_emitter, "f").off(RealtimeEvents.AttachmentFile, listener);
        };
    }
    close() {
        this.subscriber.offAll();
    }
}
_RealtimeChannel_emitter = new WeakMap();
