// @allurereport/core/dist/report.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AllureReport_instances, _AllureReport_reportName, _AllureReport_ci, _AllureReport_store, _AllureReport_readers, _AllureReport_plugins, _AllureReport_reportFiles, _AllureReport_realtimeChannel, _AllureReport_realtimeUpdateScheduler, _AllureReport_realTime, _AllureReport_hideLabels, _AllureReport_output, _AllureReport_history, _AllureReport_allureServiceClient, _AllureReport_qualityGate, _AllureReport_dump, _AllureReport_categories, _AllureReport_environments, _AllureReport_globalAttachments, _AllureReport_dumpTempDirs, _AllureReport_state, _AllureReport_executionStage, _AllureReport_publish_get, _AllureReport_runRealtimeUpdate, _AllureReport_eachPlugin, _AllureReport_getPluginState;
import console from "node:console";
import { randomUUID } from "node:crypto";
import { createWriteStream, existsSync, readFileSync } from "node:fs";
import { lstat, mkdtemp, opendir, readdir, realpath, rename, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { basename, dirname, join, resolve, sep } from "node:path";
import { promisify } from "node:util";
import { detect } from "@allurereport/ci";
import { normalizeCategoriesConfig } from "@allurereport/core-api";
import { AllureStoreDumpFiles, } from "@allurereport/plugin-api";
import { allure1, allure2, attachments, cucumberjson, junitXml, readXcResultBundle } from "@allurereport/reader";
import { PathResultFile } from "@allurereport/reader-api";
import { AllureRemoteHistory, AllureServiceClient, KnownError, UnknownError, } from "@allurereport/service";
import { generateSummary } from "@allurereport/summary";
import { glob } from "glob";
import ZipReadStream from "node-stream-zip";
import pLimit from "p-limit";
import ProgressBar from "progress";
import ZipWriteStream from "zip-stream";
import { AllureLocalHistory, createHistory } from "./history.js";
import { DefaultPluginState, PluginFiles } from "./plugin.js";
import { QualityGate } from "./qualityGate/index.js";
import { DefaultAllureStore } from "./store/store.js";
import { environmentIdentityById, environmentIdentityByName } from "./utils/environment.js";
import { RealtimeChannel } from "./utils/realtimeChannel.js";
import { RealtimeUpdateScheduler } from "./utils/realtimeUpdateScheduler.js";
import { resolveDumpAttachmentPath, UnsafeDumpPathError } from "./utils/safeDumpPath.js";
const { version } = JSON.parse(readFileSync(new URL("../package.json", import.meta.url), "utf8"));
const initRequired = "report is not initialised. Call the start() method first.";
const remoteReportParams = (ci) => {
    const repo = ci?.repoName;
    const branch = ci?.jobRunBranch;
    return repo && branch ? { repo, branch } : {};
};
const REMOTE_UPLOAD_CONCURRENCY = 50;
const REMOTE_UPLOAD_MAX_ATTEMPTS = 5;
const REMOTE_UPLOAD_MAX_SIMULTANEOUS_FAILED = 5;
const errorDetails = (err) => (err instanceof Error ? (err.stack ?? err.message) : String(err));
const isAbortError = (err) => typeof err === "object" && err !== null && "name" in err && err?.name === "AbortError";
export class AllureReport {
    constructor(opts) {
        _AllureReport_instances.add(this);
        _AllureReport_reportName.set(this, void 0);
        _AllureReport_ci.set(this, void 0);
        _AllureReport_store.set(this, void 0);
        _AllureReport_readers.set(this, void 0);
        _AllureReport_plugins.set(this, void 0);
        _AllureReport_reportFiles.set(this, void 0);
        _AllureReport_realtimeChannel.set(this, void 0);
        _AllureReport_realtimeUpdateScheduler.set(this, void 0);
        _AllureReport_realTime.set(this, void 0);
        _AllureReport_hideLabels.set(this, void 0);
        _AllureReport_output.set(this, void 0);
        _AllureReport_history.set(this, void 0);
        _AllureReport_allureServiceClient.set(this, void 0);
        _AllureReport_qualityGate.set(this, void 0);
        _AllureReport_dump.set(this, void 0);
        _AllureReport_categories.set(this, void 0);
        _AllureReport_environments.set(this, void 0);
        _AllureReport_globalAttachments.set(this, void 0);
        _AllureReport_dumpTempDirs.set(this, []);
        _AllureReport_state.set(this, void 0);
        _AllureReport_executionStage.set(this, "init");
        this.readDirectory = async (resultsDir) => {
            if (__classPrivateFieldGet(this, _AllureReport_executionStage, "f") !== "running") {
                throw new Error(initRequired);
            }
            const resultsDirPath = resolve(resultsDir);
            if (await readXcResultBundle(__classPrivateFieldGet(this, _AllureReport_store, "f"), resultsDirPath)) {
                return;
            }
            const dir = await opendir(resultsDirPath);
            try {
                for await (const dirent of dir) {
                    if (dirent.isFile()) {
                        const path = await realpath(join(dirent.parentPath ?? dirent.path, dirent.name));
                        await this.readResult(new PathResultFile(path, dirent.name));
                    }
                }
            }
            catch (e) {
                console.error("can't read directory", e);
            }
        };
        this.readFile = async (resultsFile) => {
            if (__classPrivateFieldGet(this, _AllureReport_executionStage, "f") !== "running") {
                throw new Error(initRequired);
            }
            await this.readResult(new PathResultFile(resultsFile));
        };
        this.readResult = async (data) => {
            if (__classPrivateFieldGet(this, _AllureReport_executionStage, "f") !== "running") {
                throw new Error(initRequired);
            }
            for (const reader of __classPrivateFieldGet(this, _AllureReport_readers, "f")) {
                try {
                    const processed = await reader.read(__classPrivateFieldGet(this, _AllureReport_store, "f"), data);
                    if (processed) {
                        return;
                    }
                }
                catch { }
            }
        };
        this.validate = async (params) => {
            const { trs, knownIssues, state, environment } = params;
            const qualityGateEnvironment = environment === undefined
                ? undefined
                : (environmentIdentityById(__classPrivateFieldGet(this, _AllureReport_environments, "f"), environment)?.name ??
                    environmentIdentityByName(__classPrivateFieldGet(this, _AllureReport_environments, "f"), environment)?.name ??
                    environment);
            return __classPrivateFieldGet(this, _AllureReport_qualityGate, "f").validate({
                trs: trs.filter(Boolean),
                knownIssues,
                state,
                environment: qualityGateEnvironment,
            });
        };
        this.start = async () => {
            const remoteParams = remoteReportParams(__classPrivateFieldGet(this, _AllureReport_ci, "f"));
            await __classPrivateFieldGet(this, _AllureReport_store, "f").readHistory();
            if (__classPrivateFieldGet(this, _AllureReport_executionStage, "f") === "running") {
                throw new Error("the report is already started");
            }
            if (__classPrivateFieldGet(this, _AllureReport_executionStage, "f") === "done") {
                throw new Error("the report is already stopped, the restart isn't supported at the moment");
            }
            __classPrivateFieldSet(this, _AllureReport_executionStage, "running", "f");
            const cwd = resolve(process.cwd());
            const cwdWithSep = cwd.endsWith(sep) ? cwd : `${cwd}${sep}`;
            if (__classPrivateFieldGet(this, _AllureReport_globalAttachments, "f")?.length) {
                const matchedFiles = new Set();
                for (const pattern of __classPrivateFieldGet(this, _AllureReport_globalAttachments, "f")) {
                    const files = await glob(pattern, { cwd, nodir: true, absolute: true });
                    files.forEach((filePath) => matchedFiles.add(filePath));
                }
                for (const filePath of matchedFiles) {
                    const absoluteFilePath = resolve(filePath);
                    const isInsideCwd = absoluteFilePath === cwd || absoluteFilePath.startsWith(cwdWithSep);
                    if (!isInsideCwd) {
                        continue;
                    }
                    const originalFileName = basename(absoluteFilePath);
                    __classPrivateFieldGet(this, _AllureReport_realtimeChannel, "f").dispatcher.sendGlobalAttachment(new PathResultFile(absoluteFilePath, originalFileName), originalFileName);
                }
            }
            if (__classPrivateFieldGet(this, _AllureReport_allureServiceClient, "f") && __classPrivateFieldGet(this, _AllureReport_instances, "a", _AllureReport_publish_get)) {
                const url = await __classPrivateFieldGet(this, _AllureReport_allureServiceClient, "f").createReport({
                    reportUuid: this.reportUuid,
                    reportName: __classPrivateFieldGet(this, _AllureReport_reportName, "f"),
                    ...remoteParams,
                });
                this.reportUrl = url.href;
            }
            await __classPrivateFieldGet(this, _AllureReport_eachPlugin, "f").call(this, true, async (plugin, context) => {
                await plugin.start?.(context, __classPrivateFieldGet(this, _AllureReport_store, "f"), __classPrivateFieldGet(this, _AllureReport_realtimeChannel, "f").subscriber);
            });
            if (__classPrivateFieldGet(this, _AllureReport_realTime, "f")) {
                await __classPrivateFieldGet(this, _AllureReport_runRealtimeUpdate, "f").call(this);
                __classPrivateFieldGet(this, _AllureReport_realtimeChannel, "f").onResultLikeChanged(() => {
                    __classPrivateFieldGet(this, _AllureReport_realtimeUpdateScheduler, "f").request();
                });
            }
        };
        _AllureReport_runRealtimeUpdate.set(this, async () => {
            if (__classPrivateFieldGet(this, _AllureReport_executionStage, "f") !== "running") {
                return;
            }
            await __classPrivateFieldGet(this, _AllureReport_eachPlugin, "f").call(this, false, async (plugin, context) => {
                await plugin.update?.(context, __classPrivateFieldGet(this, _AllureReport_store, "f"));
            });
        });
        this.dumpState = async () => {
            const dumpArchive = new ZipWriteStream({
                zlib: { level: 5 },
            });
            const addEntry = promisify(dumpArchive.entry.bind(dumpArchive));
            const dumpPath = `${__classPrivateFieldGet(this, _AllureReport_dump, "f")}.zip`;
            const dumpTempPath = `${dumpPath}.${randomUUID()}.tmp`;
            const dumpArchiveWriteStream = createWriteStream(dumpTempPath);
            let dumpArchiveError;
            let dumpArchiveFinished = false;
            let resolveDumpArchivePromise;
            const finishDumpArchive = (err) => {
                dumpArchiveError ?? (dumpArchiveError = err);
                if (!dumpArchiveFinished) {
                    dumpArchiveFinished = true;
                    resolveDumpArchivePromise?.();
                }
            };
            const dumpArchivePromise = new Promise((res) => {
                resolveDumpArchivePromise = res;
                dumpArchive.on("error", (err) => {
                    dumpArchiveWriteStream.destroy();
                    finishDumpArchive(err);
                });
                dumpArchiveWriteStream.on("finish", () => finishDumpArchive());
                dumpArchiveWriteStream.on("error", (err) => finishDumpArchive(err));
            });
            const errMessage = (err) => (err instanceof Error ? err.message : String(err));
            const addDumpEntry = async (data, entryName) => {
                try {
                    await addEntry(data, { name: entryName });
                    return undefined;
                }
                catch (err) {
                    return err;
                }
            };
            const addRequiredDumpEntry = async (data, entryName) => {
                const err = await addDumpEntry(data, entryName);
                if (err) {
                    throw new Error(`Failed to write dump entry "${entryName}": ${errMessage(err)}`);
                }
            };
            const addJsonDumpEntry = async (entryName, value) => {
                await addRequiredDumpEntry(Buffer.from(JSON.stringify(value)), entryName);
            };
            const dumpJsonEntries = ({ testResults, testCases, fixtures, attachments, environments, reportVariables, checkResults = [], globalAttachmentIds = [], globalErrors = [], indexAttachmentByTestResult = {}, indexTestResultByHistoryId = {}, indexTestResultByTestCase = {}, indexLatestEnvTestResultByHistoryId = {}, indexAttachmentByFixture = {}, indexFixturesByTestResult = {}, indexKnownByHistoryId = {}, qualityGateResults = [], }) => [
                [AllureStoreDumpFiles.TestResults, testResults],
                [AllureStoreDumpFiles.TestCases, testCases],
                [AllureStoreDumpFiles.Fixtures, fixtures],
                [AllureStoreDumpFiles.Attachments, attachments],
                [AllureStoreDumpFiles.CheckResults, checkResults],
                [AllureStoreDumpFiles.Environments, environments],
                [AllureStoreDumpFiles.ReportVariables, reportVariables],
                [AllureStoreDumpFiles.GlobalAttachments, globalAttachmentIds],
                [AllureStoreDumpFiles.GlobalErrors, globalErrors],
                [AllureStoreDumpFiles.IndexAttachmentsByTestResults, indexAttachmentByTestResult],
                [AllureStoreDumpFiles.IndexTestResultsByHistoryId, indexTestResultByHistoryId],
                [AllureStoreDumpFiles.IndexTestResultsByTestCase, indexTestResultByTestCase],
                [AllureStoreDumpFiles.IndexLatestEnvTestResultsByHistoryId, indexLatestEnvTestResultByHistoryId],
                [AllureStoreDumpFiles.IndexAttachmentsByFixture, indexAttachmentByFixture],
                [AllureStoreDumpFiles.IndexFixturesByTestResult, indexFixturesByTestResult],
                [AllureStoreDumpFiles.IndexKnownByHistoryId, indexKnownByHistoryId],
                [AllureStoreDumpFiles.QualityGateResults, qualityGateResults],
            ];
            let dumpError;
            dumpArchive.pipe(dumpArchiveWriteStream);
            try {
                const allAttachments = await __classPrivateFieldGet(this, _AllureReport_store, "f").allAttachments();
                for (const attachment of allAttachments) {
                    const skipAttachment = (message) => {
                        const originalFileName = attachment.originalFileName ? ` (${attachment.originalFileName})` : "";
                        __classPrivateFieldGet(this, _AllureReport_store, "f").markAttachmentMissed(attachment.id);
                        console.warn(`Skipping attachment while writing dump: ${attachment.id}${originalFileName}. ${message}`);
                    };
                    try {
                        const content = await __classPrivateFieldGet(this, _AllureReport_store, "f").attachmentContentById(attachment.id);
                        if (!content) {
                            skipAttachment("attachment content is missing");
                            continue;
                        }
                        const data = await content.asBuffer();
                        if (data === undefined) {
                            skipAttachment("attachment content is missing");
                            continue;
                        }
                        const err = await addDumpEntry(data, attachment.id);
                        if (err) {
                            skipAttachment(`failed to add attachment entry: ${errMessage(err)}`);
                        }
                    }
                    catch (err) {
                        skipAttachment(errMessage(err));
                    }
                }
                for (const [entryName, value] of dumpJsonEntries(__classPrivateFieldGet(this, _AllureReport_store, "f").dumpState())) {
                    await addJsonDumpEntry(entryName, value);
                }
            }
            catch (err) {
                dumpError = err;
            }
            finally {
                try {
                    dumpArchive.finalize();
                }
                catch (err) {
                    dumpError ?? (dumpError = err);
                    dumpArchiveWriteStream.destroy();
                    finishDumpArchive(err);
                }
                await dumpArchivePromise;
            }
            dumpError ?? (dumpError = dumpArchiveError);
            if (dumpError) {
                await rm(dumpTempPath, { force: true });
                throw dumpError;
            }
            await rename(dumpTempPath, dumpPath);
        };
        this.restoreState = async (dumps) => {
            for (const dump of dumps) {
                if (!existsSync(dump)) {
                    console.error(`Failed to restore state from "${dump}", continuing without it`);
                    console.error("Dump file does not exist");
                    continue;
                }
                try {
                    const dumpArchive = new ZipReadStream.async({
                        file: dump,
                    });
                    let restoreError;
                    try {
                        const dumpEntries = await dumpArchive.entries();
                        const dumpEntriesList = Object.entries(dumpEntries);
                        const requiredEntryData = async (entryName) => {
                            if (!dumpEntries[entryName]) {
                                throw new Error(`Missing required dump entry "${entryName}"`);
                            }
                            return dumpArchive.entryData(entryName);
                        };
                        const optionalEntryData = async (entryName) => dumpEntries[entryName] ? dumpArchive.entryData(entryName) : undefined;
                        if (!dumpEntries[AllureStoreDumpFiles.TestResults]) {
                            const nestedDumpEntries = dumpEntriesList.filter(([entryName, entry]) => !entry.isDirectory &&
                                !entryName.startsWith("__MACOSX/") &&
                                !basename(entryName).startsWith("._") &&
                                entryName.toLowerCase().endsWith(".zip"));
                            if (nestedDumpEntries.length > 0) {
                                const nestedDumpsTempDir = await mkdtemp(join(tmpdir(), `${basename(dump, ".zip")}-nested-`));
                                const nestedDumpPaths = [];
                                __classPrivateFieldGet(this, _AllureReport_dumpTempDirs, "f").push(nestedDumpsTempDir);
                                for (const [entryName] of nestedDumpEntries) {
                                    const nestedDumpPath = join(nestedDumpsTempDir, `${nestedDumpPaths.length}-${basename(entryName)}`);
                                    await writeFile(nestedDumpPath, await dumpArchive.entryData(entryName));
                                    nestedDumpPaths.push(nestedDumpPath);
                                }
                                await this.restoreState(nestedDumpPaths);
                                continue;
                            }
                        }
                        const testResultsEntry = await requiredEntryData(AllureStoreDumpFiles.TestResults);
                        const testCasesEntry = await requiredEntryData(AllureStoreDumpFiles.TestCases);
                        const fixturesEntry = await requiredEntryData(AllureStoreDumpFiles.Fixtures);
                        const attachmentsEntry = await requiredEntryData(AllureStoreDumpFiles.Attachments);
                        const checkResultsEntry = await optionalEntryData(AllureStoreDumpFiles.CheckResults);
                        const environmentsEntry = await requiredEntryData(AllureStoreDumpFiles.Environments);
                        const reportVariablesEntry = await requiredEntryData(AllureStoreDumpFiles.ReportVariables);
                        const globalAttachmentsEntry = await requiredEntryData(AllureStoreDumpFiles.GlobalAttachments);
                        const globalErrorsEntry = await requiredEntryData(AllureStoreDumpFiles.GlobalErrors);
                        const indexAttachmentsEntry = await requiredEntryData(AllureStoreDumpFiles.IndexAttachmentsByTestResults);
                        const indexTestResultsByHistoryId = await requiredEntryData(AllureStoreDumpFiles.IndexTestResultsByHistoryId);
                        const indexTestResultsByTestCaseEntry = await requiredEntryData(AllureStoreDumpFiles.IndexTestResultsByTestCase);
                        const indexLatestEnvTestResultsByHistoryIdEntry = await requiredEntryData(AllureStoreDumpFiles.IndexLatestEnvTestResultsByHistoryId);
                        const indexAttachmentsByFixtureEntry = await requiredEntryData(AllureStoreDumpFiles.IndexAttachmentsByFixture);
                        const indexFixturesByTestResultEntry = await requiredEntryData(AllureStoreDumpFiles.IndexFixturesByTestResult);
                        const indexKnownByHistoryIdEntry = await requiredEntryData(AllureStoreDumpFiles.IndexKnownByHistoryId);
                        const qualityGateResultsEntry = await requiredEntryData(AllureStoreDumpFiles.QualityGateResults);
                        const attachmentsLinks = JSON.parse(attachmentsEntry.toString("utf8"));
                        const attachmentsEntries = dumpEntriesList.reduce((acc, [entryName, entry]) => {
                            switch (entryName) {
                                case AllureStoreDumpFiles.Attachments:
                                case AllureStoreDumpFiles.CheckResults:
                                case AllureStoreDumpFiles.TestResults:
                                case AllureStoreDumpFiles.TestCases:
                                case AllureStoreDumpFiles.Fixtures:
                                case AllureStoreDumpFiles.Environments:
                                case AllureStoreDumpFiles.ReportVariables:
                                case AllureStoreDumpFiles.GlobalAttachments:
                                case AllureStoreDumpFiles.GlobalErrors:
                                case AllureStoreDumpFiles.IndexAttachmentsByTestResults:
                                case AllureStoreDumpFiles.IndexTestResultsByHistoryId:
                                case AllureStoreDumpFiles.IndexTestResultsByTestCase:
                                case AllureStoreDumpFiles.IndexLatestEnvTestResultsByHistoryId:
                                case AllureStoreDumpFiles.IndexAttachmentsByFixture:
                                case AllureStoreDumpFiles.IndexFixturesByTestResult:
                                case AllureStoreDumpFiles.IndexKnownByHistoryId:
                                case AllureStoreDumpFiles.QualityGateResults:
                                    return acc;
                                default:
                                    if (entry.isDirectory || !attachmentsLinks[entryName] || attachmentsLinks[entryName].missed) {
                                        return acc;
                                    }
                                    return Object.assign(acc, {
                                        [entryName]: entry,
                                    });
                            }
                        }, {});
                        const dumpState = {
                            testResults: JSON.parse(testResultsEntry.toString("utf8")),
                            testCases: JSON.parse(testCasesEntry.toString("utf8")),
                            fixtures: JSON.parse(fixturesEntry.toString("utf8")),
                            attachments: attachmentsLinks,
                            checkResults: checkResultsEntry ? JSON.parse(checkResultsEntry.toString("utf8")) : [],
                            environments: JSON.parse(environmentsEntry.toString("utf8")),
                            reportVariables: JSON.parse(reportVariablesEntry.toString("utf8")),
                            globalAttachmentIds: JSON.parse(globalAttachmentsEntry.toString("utf8")),
                            globalErrors: JSON.parse(globalErrorsEntry.toString("utf8")),
                            indexAttachmentByTestResult: JSON.parse(indexAttachmentsEntry.toString("utf8")),
                            indexTestResultByHistoryId: JSON.parse(indexTestResultsByHistoryId.toString("utf8")),
                            indexTestResultByTestCase: JSON.parse(indexTestResultsByTestCaseEntry.toString("utf8")),
                            indexLatestEnvTestResultByHistoryId: JSON.parse(indexLatestEnvTestResultsByHistoryIdEntry.toString("utf8")),
                            indexAttachmentByFixture: JSON.parse(indexAttachmentsByFixtureEntry.toString("utf8")),
                            indexFixturesByTestResult: JSON.parse(indexFixturesByTestResultEntry.toString("utf8")),
                            indexKnownByHistoryId: JSON.parse(indexKnownByHistoryIdEntry.toString("utf8")),
                            qualityGateResults: JSON.parse(qualityGateResultsEntry.toString("utf8")),
                        };
                        const dumpTempDir = await mkdtemp(join(tmpdir(), basename(dump, ".zip")));
                        const resultsAttachments = {};
                        __classPrivateFieldGet(this, _AllureReport_dumpTempDirs, "f").push(dumpTempDir);
                        try {
                            for (const [attachmentId] of Object.entries(attachmentsEntries)) {
                                const attachmentContentEntry = await dumpArchive.entryData(attachmentId);
                                const attachmentFilePath = resolveDumpAttachmentPath(dumpTempDir, attachmentId);
                                await writeFile(attachmentFilePath, attachmentContentEntry);
                                resultsAttachments[attachmentId] = new PathResultFile(attachmentFilePath, attachmentId);
                            }
                        }
                        catch (err) {
                            if (err instanceof UnsafeDumpPathError) {
                                console.error(`Cannot restore dump from "${dump}": the archive lists attachment paths that would write outside the extract directory (unsafe zip paths such as "../" or absolute names).`);
                                console.error(err.message);
                                console.error("Only use dump archives produced by this tool; do not load untrusted or third-party --dump zip files.");
                                throw err;
                            }
                            console.error(`Can't restore attachment contents from "${dump}", continuing without them`);
                            console.error(errorDetails(err));
                        }
                        await __classPrivateFieldGet(this, _AllureReport_store, "f").restoreState(dumpState, resultsAttachments);
                        console.info(`Successfully restored state from "${dump}"`);
                    }
                    catch (err) {
                        restoreError = err;
                        throw err;
                    }
                    finally {
                        try {
                            await dumpArchive.close();
                        }
                        catch (err) {
                            if (!restoreError) {
                                console.error(`Failed to close dump archive "${dump}"`);
                                console.error(errorDetails(err));
                            }
                        }
                    }
                }
                catch (err) {
                    console.error(`Failed to restore state from "${dump}", continuing without it`);
                    console.error(errorDetails(err));
                }
            }
        };
        this.done = async () => {
            const summaries = [];
            const remoteHrefs = new Set();
            const remoteHrefsByPluginId = {};
            const cancelledPluginsIds = new Set();
            let remoteCleanupFailed = false;
            const getSuccessfulPublishedPlugins = () => __classPrivateFieldGet(this, _AllureReport_plugins, "f").filter(({ enabled, id, options }) => enabled && !!options?.publish && !cancelledPluginsIds.has(id));
            if (__classPrivateFieldGet(this, _AllureReport_executionStage, "f") !== "running") {
                throw new Error(initRequired);
            }
            const testResults = await __classPrivateFieldGet(this, _AllureReport_store, "f").allTestResults();
            const testCases = await __classPrivateFieldGet(this, _AllureReport_store, "f").allTestCases();
            const historyDataPoint = createHistory(this.reportUuid, __classPrivateFieldGet(this, _AllureReport_reportName, "f"), testCases, testResults, this.reportUrl);
            __classPrivateFieldGet(this, _AllureReport_realtimeChannel, "f").close();
            try {
                await __classPrivateFieldGet(this, _AllureReport_realtimeUpdateScheduler, "f").close();
            }
            catch (e) {
                console.error("realtime update failed during shutdown", e);
            }
            __classPrivateFieldSet(this, _AllureReport_executionStage, "done", "f");
            if (__classPrivateFieldGet(this, _AllureReport_dump, "f")) {
                await this.dumpState();
                return;
            }
            await __classPrivateFieldGet(this, _AllureReport_eachPlugin, "f").call(this, false, async (plugin, context) => {
                await plugin.done?.(context, __classPrivateFieldGet(this, _AllureReport_store, "f"));
            });
            await __classPrivateFieldGet(this, _AllureReport_eachPlugin, "f").call(this, false, async (plugin, context) => {
                if (__classPrivateFieldGet(this, _AllureReport_allureServiceClient, "f") && context.publish) {
                    const pluginFiles = (await context.state.get("files")) ?? {};
                    const pluginFilesEntries = Object.entries(pluginFiles);
                    const progressBar = pluginFilesEntries?.length > 0
                        ? new ProgressBar(`Publishing "${context.id}" report [:bar] :current/:total`, {
                            total: pluginFilesEntries.length,
                            width: 20,
                        })
                        : undefined;
                    const limitFn = pLimit(REMOTE_UPLOAD_CONCURRENCY);
                    const uploadAbortController = new AbortController();
                    const failedUploads = new Set();
                    let terminalUploadError;
                    let remoteReportDeleted = false;
                    let fns = [];
                    const cleanupFailedPluginUpload = async (err) => {
                        if (remoteReportDeleted) {
                            return;
                        }
                        remoteReportDeleted = true;
                        cancelledPluginsIds.add(context.id);
                        uploadAbortController.abort();
                        await Promise.allSettled(fns);
                        const pluginRemoteHref = remoteHrefsByPluginId[context.id];
                        if (pluginRemoteHref) {
                            remoteHrefs.delete(pluginRemoteHref);
                            delete remoteHrefsByPluginId[context.id];
                        }
                        try {
                            await __classPrivateFieldGet(this, _AllureReport_allureServiceClient, "f").deleteReport({
                                reportUuid: this.reportUuid,
                                pluginId: context.id,
                            });
                        }
                        catch (cleanupErr) {
                            remoteCleanupFailed = true;
                            console.error(`Plugin "${context.id}" upload cleanup has failed, the remote report won't be completed`);
                            console.error(cleanupErr);
                        }
                        console.error(`Plugin "${context.id}" upload has failed, the plugin won't be published`);
                        console.error(err);
                    };
                    fns = pluginFilesEntries.map(([filename, filepath]) => limitFn(async () => {
                        if (cancelledPluginsIds.has(context.id) || uploadAbortController.signal.aborted) {
                            return;
                        }
                        for (let attempt = 1; attempt <= REMOTE_UPLOAD_MAX_ATTEMPTS; attempt++) {
                            if (cancelledPluginsIds.has(context.id) || uploadAbortController.signal.aborted) {
                                return;
                            }
                            try {
                                if (/^(data|widgets|index\.html$|summary\.json$)/.test(filename)) {
                                    const uploadedFileUrl = await __classPrivateFieldGet(this, _AllureReport_allureServiceClient, "f").addReportFile({
                                        reportUuid: this.reportUuid,
                                        pluginId: context.id,
                                        filename,
                                        filepath,
                                        signal: uploadAbortController.signal,
                                    });
                                    if (cancelledPluginsIds.has(context.id) || uploadAbortController.signal.aborted) {
                                        return;
                                    }
                                    if (filename === "index.html") {
                                        remoteHrefsByPluginId[context.id] = uploadedFileUrl;
                                        remoteHrefs.add(uploadedFileUrl);
                                    }
                                }
                                else {
                                    await __classPrivateFieldGet(this, _AllureReport_allureServiceClient, "f").addReportAsset({
                                        filename,
                                        filepath,
                                        signal: uploadAbortController.signal,
                                    });
                                }
                                failedUploads.delete(filename);
                                progressBar?.tick?.();
                                return;
                            }
                            catch (err) {
                                if (cancelledPluginsIds.has(context.id) || uploadAbortController.signal.aborted) {
                                    return;
                                }
                                if (isAbortError(err)) {
                                    throw err;
                                }
                                failedUploads.add(filename);
                                if (failedUploads.size > REMOTE_UPLOAD_MAX_SIMULTANEOUS_FAILED ||
                                    attempt >= REMOTE_UPLOAD_MAX_ATTEMPTS) {
                                    terminalUploadError ?? (terminalUploadError = err);
                                    throw terminalUploadError;
                                }
                            }
                        }
                    }));
                    progressBar?.render?.();
                    try {
                        await Promise.all(fns);
                    }
                    catch (err) {
                        await cleanupFailedPluginUpload(terminalUploadError ?? err);
                    }
                }
                const summary = await plugin?.info?.(context, __classPrivateFieldGet(this, _AllureReport_store, "f"));
                if (!summary) {
                    return;
                }
                summary.pluginId = context.id;
                summary.pullRequestHref = __classPrivateFieldGet(this, _AllureReport_ci, "f")?.pullRequestUrl;
                summary.jobHref = __classPrivateFieldGet(this, _AllureReport_ci, "f")?.jobRunUrl;
                if (context.publish && !cancelledPluginsIds.has(context.id)) {
                    summary.remoteHref =
                        remoteHrefsByPluginId[context.id] ?? (this.reportUrl ? `${this.reportUrl}/${context.id}/` : undefined);
                    if (summary.remoteHref) {
                        remoteHrefs.add(summary.remoteHref);
                    }
                }
                summaries.push({
                    ...summary,
                    href: `${context.id}/`,
                });
                await context.reportFiles.addFile("summary.json", Buffer.from(JSON.stringify(summary)));
            });
            if (summaries.length > 1) {
                const summaryPath = await generateSummary(__classPrivateFieldGet(this, _AllureReport_output, "f"), summaries);
                const publishedReports = getSuccessfulPublishedPlugins();
                if (__classPrivateFieldGet(this, _AllureReport_instances, "a", _AllureReport_publish_get) && summaryPath && publishedReports.length > 1) {
                    await __classPrivateFieldGet(this, _AllureReport_allureServiceClient, "f")?.addReportFile({
                        reportUuid: this.reportUuid,
                        filename: "index.html",
                        filepath: summaryPath,
                    });
                }
            }
            const publishedReports = getSuccessfulPublishedPlugins();
            if (__classPrivateFieldGet(this, _AllureReport_instances, "a", _AllureReport_publish_get) && !remoteCleanupFailed && publishedReports.length > 0) {
                await __classPrivateFieldGet(this, _AllureReport_allureServiceClient, "f")?.completeReport({
                    reportUuid: this.reportUuid,
                    historyPoint: historyDataPoint,
                });
            }
            let outputDirFiles = [];
            try {
                outputDirFiles = await readdir(__classPrivateFieldGet(this, _AllureReport_output, "f"));
            }
            catch { }
            if (outputDirFiles.length === 0) {
                return;
            }
            const reportPath = join(__classPrivateFieldGet(this, _AllureReport_output, "f"), outputDirFiles[0]);
            const reportStats = await lstat(reportPath);
            const outputEntriesStats = await Promise.all(outputDirFiles.map((file) => lstat(join(__classPrivateFieldGet(this, _AllureReport_output, "f"), file))));
            const outputDirectoryEntries = outputEntriesStats.filter((entry) => entry.isDirectory());
            if (reportStats.isDirectory() && outputDirectoryEntries.length === 1) {
                const reportContent = await readdir(reportPath);
                for (const entry of reportContent) {
                    const currentFilePath = join(reportPath, entry);
                    const newFilePath = resolve(dirname(currentFilePath), "..", entry);
                    await rename(currentFilePath, newFilePath);
                }
                await rm(reportPath, { recursive: true });
            }
            for (const dir of __classPrivateFieldGet(this, _AllureReport_dumpTempDirs, "f")) {
                try {
                    await rm(dir, { recursive: true });
                }
                catch { }
            }
            if (__classPrivateFieldGet(this, _AllureReport_history, "f")) {
                try {
                    await __classPrivateFieldGet(this, _AllureReport_store, "f").appendHistory(historyDataPoint);
                }
                catch (err) {
                    if (err instanceof KnownError) {
                        console.error("Failed to append history", err.message);
                    }
                    else if (err instanceof UnknownError) {
                        console.error("Failed to append history due to unexpected error", err.message);
                    }
                    else {
                        throw err;
                    }
                }
            }
            if (!remoteCleanupFailed && remoteHrefs.size > 0) {
                console.info("Next reports have been published:");
                remoteHrefs.forEach((href) => {
                    console.info(`- ${href}`);
                });
            }
            if (!__classPrivateFieldGet(this, _AllureReport_qualityGate, "f")) {
                return;
            }
            const qualityGateResults = await __classPrivateFieldGet(this, _AllureReport_store, "f").qualityGateResultsByEnv();
            await writeFile(join(__classPrivateFieldGet(this, _AllureReport_output, "f"), "quality-gate.json"), JSON.stringify(qualityGateResults));
        };
        _AllureReport_eachPlugin.set(this, async (initState, consumer) => {
            if (initState) {
                __classPrivateFieldSet(this, _AllureReport_state, {}, "f");
            }
            for (const { enabled, id, plugin, options } of __classPrivateFieldGet(this, _AllureReport_plugins, "f")) {
                if (!enabled) {
                    continue;
                }
                const pluginState = __classPrivateFieldGet(this, _AllureReport_instances, "m", _AllureReport_getPluginState).call(this, initState, id);
                if (!pluginState) {
                    console.error("plugin error: state is empty");
                    continue;
                }
                if (initState) {
                    await pluginState.set("files", {});
                }
                const pluginFiles = new PluginFiles(__classPrivateFieldGet(this, _AllureReport_reportFiles, "f"), id, async (key, filepath) => {
                    const currentPluginState = __classPrivateFieldGet(this, _AllureReport_instances, "m", _AllureReport_getPluginState).call(this, false, id);
                    const files = await currentPluginState?.get("files");
                    if (!files) {
                        return;
                    }
                    files[key] = filepath;
                });
                const pluginContext = {
                    id,
                    publish: !!options?.publish,
                    allureVersion: version,
                    reportUuid: this.reportUuid,
                    reportName: __classPrivateFieldGet(this, _AllureReport_reportName, "f"),
                    hideLabels: __classPrivateFieldGet(this, _AllureReport_hideLabels, "f"),
                    state: pluginState,
                    reportFiles: pluginFiles,
                    reportUrl: this.reportUrl,
                    output: __classPrivateFieldGet(this, _AllureReport_output, "f"),
                    ci: __classPrivateFieldGet(this, _AllureReport_ci, "f"),
                    categories: __classPrivateFieldGet(this, _AllureReport_categories, "f"),
                    history: __classPrivateFieldGet(this, _AllureReport_history, "f"),
                };
                try {
                    await consumer.call(this, plugin, pluginContext);
                    if (initState) {
                        __classPrivateFieldGet(this, _AllureReport_state, "f")[id] = pluginState;
                    }
                }
                catch (e) {
                    console.error(`plugin ${id} error`, e);
                }
            }
        });
        const { name, readers = [allure1, allure2, cucumberjson, junitXml, attachments], plugins = [], known, reportFiles, realTime, historyPath, historyLimit, defaultLabels = {}, variables = {}, environment, allowedEnvironments, environments, output, hideLabels, qualityGate, dump, categories, allureService: allureServiceConfig, globalAttachments, } = opts;
        if (allureServiceConfig?.accessToken) {
            __classPrivateFieldSet(this, _AllureReport_allureServiceClient, new AllureServiceClient(allureServiceConfig), "f");
        }
        this.reportUuid = randomUUID();
        __classPrivateFieldSet(this, _AllureReport_ci, detect(), "f");
        const reportTitleSuffix = __classPrivateFieldGet(this, _AllureReport_ci, "f")?.pullRequestName ?? __classPrivateFieldGet(this, _AllureReport_ci, "f")?.jobRunName;
        __classPrivateFieldSet(this, _AllureReport_reportName, [name, reportTitleSuffix].filter(Boolean).join(" – "), "f");
        __classPrivateFieldSet(this, _AllureReport_realtimeChannel, new RealtimeChannel(), "f");
        __classPrivateFieldSet(this, _AllureReport_realtimeUpdateScheduler, new RealtimeUpdateScheduler(__classPrivateFieldGet(this, _AllureReport_runRealtimeUpdate, "f")), "f");
        __classPrivateFieldSet(this, _AllureReport_realTime, realTime, "f");
        __classPrivateFieldSet(this, _AllureReport_dump, dump, "f");
        __classPrivateFieldSet(this, _AllureReport_hideLabels, hideLabels, "f");
        __classPrivateFieldSet(this, _AllureReport_environments, environments ?? {}, "f");
        __classPrivateFieldSet(this, _AllureReport_globalAttachments, globalAttachments, "f");
        if (qualityGate) {
            __classPrivateFieldSet(this, _AllureReport_qualityGate, new QualityGate(qualityGate), "f");
        }
        __classPrivateFieldSet(this, _AllureReport_categories, normalizeCategoriesConfig(categories), "f");
        if (__classPrivateFieldGet(this, _AllureReport_allureServiceClient, "f")) {
            __classPrivateFieldSet(this, _AllureReport_history, new AllureRemoteHistory({
                limit: historyLimit,
                ...remoteReportParams(__classPrivateFieldGet(this, _AllureReport_ci, "f")),
                allureServiceClient: __classPrivateFieldGet(this, _AllureReport_allureServiceClient, "f"),
            }), "f");
        }
        else if (historyPath) {
            __classPrivateFieldSet(this, _AllureReport_history, new AllureLocalHistory({
                limit: historyLimit,
                historyPath,
            }), "f");
        }
        __classPrivateFieldSet(this, _AllureReport_store, new DefaultAllureStore({
            realtimeSubscriber: __classPrivateFieldGet(this, _AllureReport_realtimeChannel, "f").subscriber,
            realtimeDispatcher: __classPrivateFieldGet(this, _AllureReport_realtimeChannel, "f").dispatcher,
            reportVariables: variables,
            environmentsConfig: environments,
            history: __classPrivateFieldGet(this, _AllureReport_history, "f"),
            known,
            defaultLabels,
            environment,
            allowedEnvironments,
        }), "f");
        __classPrivateFieldSet(this, _AllureReport_readers, [...readers], "f");
        __classPrivateFieldSet(this, _AllureReport_plugins, [...plugins], "f");
        __classPrivateFieldSet(this, _AllureReport_reportFiles, reportFiles, "f");
        __classPrivateFieldSet(this, _AllureReport_output, output, "f");
    }
    get hasQualityGate() {
        return !!__classPrivateFieldGet(this, _AllureReport_qualityGate, "f");
    }
    get store() {
        return __classPrivateFieldGet(this, _AllureReport_store, "f");
    }
    get realtimeSubscriber() {
        return __classPrivateFieldGet(this, _AllureReport_realtimeChannel, "f").subscriber;
    }
    get realtimeDispatcher() {
        return __classPrivateFieldGet(this, _AllureReport_realtimeChannel, "f").dispatcher;
    }
}
_AllureReport_reportName = new WeakMap(), _AllureReport_ci = new WeakMap(), _AllureReport_store = new WeakMap(), _AllureReport_readers = new WeakMap(), _AllureReport_plugins = new WeakMap(), _AllureReport_reportFiles = new WeakMap(), _AllureReport_realtimeChannel = new WeakMap(), _AllureReport_realtimeUpdateScheduler = new WeakMap(), _AllureReport_realTime = new WeakMap(), _AllureReport_hideLabels = new WeakMap(), _AllureReport_output = new WeakMap(), _AllureReport_history = new WeakMap(), _AllureReport_allureServiceClient = new WeakMap(), _AllureReport_qualityGate = new WeakMap(), _AllureReport_dump = new WeakMap(), _AllureReport_categories = new WeakMap(), _AllureReport_environments = new WeakMap(), _AllureReport_globalAttachments = new WeakMap(), _AllureReport_dumpTempDirs = new WeakMap(), _AllureReport_state = new WeakMap(), _AllureReport_executionStage = new WeakMap(), _AllureReport_runRealtimeUpdate = new WeakMap(), _AllureReport_eachPlugin = new WeakMap(), _AllureReport_instances = new WeakSet(), _AllureReport_publish_get = function _AllureReport_publish_get() {
    return __classPrivateFieldGet(this, _AllureReport_plugins, "f").some(({ enabled, options }) => enabled && options.publish);
}, _AllureReport_getPluginState = function _AllureReport_getPluginState(init, id) {
    return init ? new DefaultPluginState({}) : __classPrivateFieldGet(this, _AllureReport_state, "f")?.[id];
};
