// @allurereport/service/dist/service.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AllureServiceClient_client, _AllureServiceClient_url;
import { readFile } from "node:fs/promises";
import { join as joinPosix } from "node:path/posix";
import { createServiceHttpClient } from "./utils/http.js";
import { parseServiceToken } from "./utils/token.js";
const UPLOAD_CONTENT_TYPE = "application/octet-stream";
const createUploadBlob = (content) => new Blob([content], { type: UPLOAD_CONTENT_TYPE });
const createReportUrl = (baseUrl, reportUuid) => `${baseUrl}/${reportUuid}`;
const createReportFileUrl = (baseUrl, reportUuid, reportFilename) => `${baseUrl}/${joinPosix(reportUuid, reportFilename)}`;
export class AllureServiceClient {
    constructor(config) {
        this.config = config;
        _AllureServiceClient_client.set(this, void 0);
        _AllureServiceClient_url.set(this, void 0);
        if (!config?.accessToken) {
            throw new Error("Allure service access token is required");
        }
        const { url } = parseServiceToken(config.accessToken);
        __classPrivateFieldSet(this, _AllureServiceClient_url, url.replace(/\/$/, ""), "f");
        __classPrivateFieldSet(this, _AllureServiceClient_client, createServiceHttpClient(__classPrivateFieldGet(this, _AllureServiceClient_url, "f"), config.accessToken), "f");
    }
    async downloadHistory(payload) {
        const { repo, branch, limit } = payload ?? {};
        const { history } = await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").get("/api/history", {
            params: {
                limit: limit ? encodeURIComponent(limit) : undefined,
                repo: repo ? encodeURIComponent(repo) : undefined,
                branch: branch ? encodeURIComponent(branch) : undefined,
            },
        });
        return history;
    }
    async createReport(payload) {
        const { reportName, reportUuid, repo, branch } = payload;
        const { url } = await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post("/api/reports", {
            body: {
                reportName,
                reportUuid,
                repo,
                branch,
            },
        });
        return new URL(url, __classPrivateFieldGet(this, _AllureServiceClient_url, "f"));
    }
    async completeReport(payload) {
        const { reportUuid, historyPoint } = payload;
        const completedHistoryPoint = {
            ...historyPoint,
            url: createReportUrl(__classPrivateFieldGet(this, _AllureServiceClient_url, "f"), reportUuid),
        };
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post(`/api/reports/${reportUuid}/complete`, {
            body: {
                historyPoint: completedHistoryPoint,
            },
        });
    }
    async deleteReport(payload) {
        const { reportUuid, pluginId = "" } = payload;
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post(`/api/report/${reportUuid}/delete`, {
            body: {
                pluginId,
            },
        });
    }
    async addReportAsset(payload) {
        const { filename, file, filepath, signal } = payload;
        if (!file && !filepath) {
            throw new Error("File or filepath is required");
        }
        let content = file;
        if (!content) {
            content = signal ? await readFile(filepath, { signal }) : await readFile(filepath);
        }
        const form = new FormData();
        form.set("filename", filename);
        form.set("file", createUploadBlob(content), filename);
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post("/api/assets/upload", {
            body: form,
            headers: {
                "Content-Type": "multipart/form-data",
            },
            ...(signal ? { signal } : {}),
        });
    }
    async addReportFile(payload) {
        const { reportUuid, filename, file, filepath, pluginId, signal } = payload;
        const reportFilename = pluginId ? joinPosix(pluginId, filename) : filename;
        if (!file && !filepath) {
            throw new Error("File or filepath is required");
        }
        let content = file;
        if (!content) {
            content = signal ? await readFile(filepath, { signal }) : await readFile(filepath);
        }
        const form = new FormData();
        form.set("filename", reportFilename);
        form.set("file", createUploadBlob(content), reportFilename);
        await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post(`/api/reports/${reportUuid}/upload`, {
            body: form,
            headers: {
                "Content-Type": "multipart/form-data",
            },
            ...(signal ? { signal } : {}),
        });
        return createReportFileUrl(__classPrivateFieldGet(this, _AllureServiceClient_url, "f"), reportUuid, reportFilename);
    }
}
_AllureServiceClient_client = new WeakMap(), _AllureServiceClient_url = new WeakMap();
